### PR TITLE
Add insecure flag to allow bypassing SSL hostname verification and cert checks

### PR DESCRIPTION
**Pull request by MattLud**
Initial proposal - Let me know what sort of unit tests you'd like to see or if you want the whitespace changes reverted.

### CODE REVIEW

File: src/main/java/hudson/remoting/Engine.java [hunk 1]
```diff
@@ -154,9 +152,15 @@ public void run() {
      */
     private boolean keepAlive = true;
 
-    
-    
-    
+
+    public boolean isInsecure() {
```
Comment (oleg-nenashev) [inline src/main/java/hudson/remoting/Engine.java]: Methods should not be listed before fields. It will confuse other contributors

File: src/main/java/org/jenkinsci/remoting/engine/JnlpAgentEndpointResolver.java [hunk 2]
```diff
@@ -80,6 +89,16 @@
 
     private String tunnel;
 
+    private boolean insecure;
+
+    public boolean isInsecure() {
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/remoting/engine/JnlpAgentEndpointResolver.java]: same

File: src/main/java/org/jenkinsci/remoting/engine/JnlpAgentEndpointResolver.java [hunk 3]
```diff
@@ -141,7 +160,7 @@ public void setTunnel(String tunnel) {
     public JnlpAgentEndpoint resolve() throws IOException {
         IOException firstError = null;
         for (String jenkinsUrl : jenkinsUrls) {
-
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/remoting/engine/JnlpAgentEndpointResolver.java]: Please do not do unrelated whitespace changes. It may cause undesired merge conflicts

File: src/main/java/hudson/remoting/Engine.java (hunk 1 above)
Comment (oleg-nenashev) [reply src/main/java/hudson/remoting/Engine.java]: New API methods also should be documented

**Comment by MattLud** [comment]
Ok - repushed with amended commit.

File: src/main/java/hudson/remoting/jnlp/Main.java [hunk 4]
```diff
@@ -92,6 +89,10 @@
             usage="If the connection ends, don't retry and just exit.")
     public boolean noReconnect = false;
 
+    @Option(name="-insecure",
+        usage="Ignore SSL validation errors - use as a last resort only.")
```
Comment (Wadeck) [inline src/main/java/hudson/remoting/jnlp/Main.java]: Exactly the type of comment I would like to see more often on such option :) :+1:

File: src/main/java/org/jenkinsci/remoting/engine/JnlpAgentEndpointResolver.java [hunk 5]
```diff
@@ -137,6 +148,25 @@ public void setTunnel(String tunnel) {
         this.tunnel = tunnel;
     }
 
+    /**
+     *  Determine if certificate checking should be ignored for JNLP endpoint
+     *
+     * @return if insecure, endpoint check is ignored
+     */
+
+    public boolean isInsecure() {
```
Comment (Wadeck) [inline src/main/java/org/jenkinsci/remoting/engine/JnlpAgentEndpointResolver.java]: Weird line break +1

File: src/main/java/org/jenkinsci/remoting/engine/JnlpAgentEndpointResolver.java [hunk 6]
```diff
@@ -407,8 +436,54 @@ static URLConnection openURLConnection(URL url, String credentials, String proxy
             String encoding = Base64.encode(proxyCredentials.getBytes("UTF-8"));
             con.setRequestProperty("Proxy-Authorization", "Basic " + encoding);
         }
-        if (con instanceof HttpsURLConnection && sslSocketFactory != null) {
+
+        if (insecure && con instanceof HttpsURLConnection) {
+            System.out.println(String.format("Insecure Status: %s", insecure));
```
Comment (Wadeck) [inline src/main/java/org/jenkinsci/remoting/engine/JnlpAgentEndpointResolver.java]: Better to use a logger instead of System.out.println.

File: src/main/java/org/jenkinsci/remoting/engine/JnlpAgentEndpointResolver.java [hunk 7]
```diff
@@ -407,8 +436,54 @@ static URLConnection openURLConnection(URL url, String credentials, String proxy
...
+
+                HostnameVerifier allHostsValid = new HostnameVerifier() {
+                    @Override
+                    public boolean verify(String hostname, SSLSession session) {
+                        return true;
+                    }
+                };
+
+                ((HttpsURLConnection) con).setHostnameVerifier(allHostsValid);
+            ((HttpsURLConnection) con).setSSLSocketFactory(sslSocketFactory);
```
Comment (Wadeck) [inline src/main/java/org/jenkinsci/remoting/engine/JnlpAgentEndpointResolver.java]: Double casting could be avoided by using a local variable just above like
```
HttpsURLConnection httpsCon = (HttpsURLConnection) con;
httpsCon.setHostnameVerifier(allHostsValid);
httpsCon.setSSLSocketFactory(sslSocketFactory);
```

File: src/main/java/org/jenkinsci/remoting/engine/JnlpAgentEndpointResolver.java [hunk 8]
```diff
@@ -407,8 +436,54 @@ static URLConnection openURLConnection(URL url, String credentials, String proxy
...
+        }
+        else if (con instanceof HttpsURLConnection && sslSocketFactory != null) {
             ((HttpsURLConnection) con).setSSLSocketFactory(sslSocketFactory);
+            HostnameVerifier allHostsValid = new HostnameVerifier() {
+                @Override
+                public boolean verify(String hostname, SSLSession session) {
+                    return true;
+                }
+            };
+            ((HttpsURLConnection) con).setHostnameVerifier(allHostsValid);
```
Comment (Wadeck) [inline src/main/java/org/jenkinsci/remoting/engine/JnlpAgentEndpointResolver.java]: Ditto double cast

File: src/main/java/org/jenkinsci/remoting/engine/JnlpAgentEndpointResolver.java (hunk 6 above)
Comment (oleg-nenashev) [reply src/main/java/org/jenkinsci/remoting/engine/JnlpAgentEndpointResolver.java]: It requires a wider refactoring, the option is being used in other places

**Comment by MattLud** [comment]
Ok - is there any change you need from me right now?

**Comment by oleg-nenashev** [comment]
@MattLud Created https://github.com/jenkinsci/remoting/pull/210
